aws/resource_aws_traffic_mirror_filter_rule.go

```diff
@@ -0,0 +1,369 @@
+
+	if d.HasChange("description") {
+		_, n := d.GetChange("description")
+		if n != "" {
+			input.Description = aws.String(n.(string))
+		} else {
+			removeFields = append(removeFields, aws.String("description"))
+		}
+	}
+
+	if d.HasChange("destination_cidr_block") {
+		input.DestinationCidrBlock = aws.String(d.Get("destination_cidr_block").(string))
+	}
+
+	if d.HasChange("source_cidr_block") {
+		input.SourceCidrBlock = aws.String(d.Get("source_cidr_block").(string))
+	}
+
+	if d.HasChange("destination_port_range") {
+		_, v := d.GetChange("destination_port_range")
+		n := v.([]interface{})
+		if 0 == len(n) {
+			removeFields = append(removeFields, aws.String("destination-port-range"))
+		} else {
+			//Modify request that adds port range seems to fail if protocol is not set in the request
+			input.Protocol = aws.Int64(int64(d.Get("protocol").(int)))
+			input.SetDestinationPortRange(buildTrafficMirrorPortRangeRequest(n))
+		}
+	}
+
+	if d.HasChange("source_port_range") {
+		_, v := d.GetChange("source_port_range")
+		n := v.([]interface{})
+		if 0 == len(n) {
+			removeFields = append(removeFields, aws.String("source-port-range"))
+		} else {
+			//Modify request that adds port range seems to fail if protocol is not set in the request
+			input.Protocol = aws.Int64(int64(d.Get("protocol").(int)))
+			input.SetSourcePortRange(buildTrafficMirrorPortRangeRequest(n))
+		}
+	}
+
+	if d.HasChange("rule_action") {
+		input.RuleAction = aws.String(d.Get("rule_action").(string))
+	}
+
+	if d.HasChange("rule_number") {
+		input.RuleNumber = aws.Int64(int64(d.Get("rule_action").(int)))
+	}
+
+	if d.HasChange("traffic_direction") {
+		input.TrafficDirection = aws.String(d.Get("traffic_direction").(string))
+	}
+
+	if len(removeFields) > 0 {
+		input.SetRemoveFields(removeFields)
+	}
+
+	_, err := conn.ModifyTrafficMirrorFilterRule(input)
+	if err != nil {
+		return fmt.Errorf("Error modifying rule %v", ruleId)
+	}
+
+	return resourceAwsTrafficMirrorFilterRuleRead(d, meta)
+}
+
+func resourceAwsTrafficMirrorFilterRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	ruleId := d.Id()
+	input := &ec2.DeleteTrafficMirrorFilterRuleInput{
+		TrafficMirrorFilterRuleId: &ruleId,
+	}
+
+	_, err := conn.DeleteTrafficMirrorFilterRule(input)
+	if err != nil {
+		return fmt.Errorf("Error deleting traffic mirror filter rule %v", ruleId)
+	}
+
+	return nil
+}
+
+func buildTrafficMirrorPortRangeRequest(p []interface{}) (out *ec2.TrafficMirrorPortRangeRequest) {
+	portSchema := p[0].(map[string]interface{})
+
+	portRange := ec2.TrafficMirrorPortRangeRequest{}
+	if v, ok := portSchema["from_port"]; ok {
+		portRange.FromPort = aws.Int64(int64(v.(int)))
+		out = &portRange
+	}
+
+	if v, ok := portSchema["to_port"]; ok {
+		portRange.ToPort = aws.Int64(int64(v.(int)))
+		out = &portRange
+	}
+
+	return out
+}
+
+func populateTrafficMirrorFilterRuleResource(d *schema.ResourceData, rule *ec2.TrafficMirrorFilterRule) error {
```

Comment:
Since this is only used once, let's remove the indirection and include this directly in the Read function. 👍